driver/CabalWrapper.hs

```diff
@@ -0,0 +1,44 @@
+module Main (main) where
+
+import GHC.Paths (ghc_pkg)
+import System.Environment (getArgs, getEnvironment, getExecutablePath, lookupEnv)
+import System.Exit (exitWith)
+import qualified System.Process as P
+import Test.DocTest (doctest)
+import Prelude
+
+main :: IO ()
+main = getArgs >>= dispatch
+
+{- | Check if @$CABAL@ is defined to see if used as cabal external command or as
+a compiler, then run corresponding actions.
+-}
+dispatch :: [String] -> IO ()
+dispatch args = do
+    mCabal <- lookupEnv cabalEnvVariable
+    case mCabal of
+        Nothing -> doctest args
+        -- Drop first arg
+        -- See: https://cabal.readthedocs.io/en/latest/external-commands.html
+        Just cabal -> cabalRepl cabal (drop 1 args)
+
+-- | Run as external cabal command: run cabal repl with relevant compiler options
+cabalRepl :: String -> [String] -> IO ()
+cabalRepl cabal args = do
+    exePath <- getExecutablePath
+    -- Remove `cabalEnvVariable` from the environment, else we will get an infinite loop
+    cur_env <- getEnvironment
+    let new_env = filter ((/= cabalEnvVariable) . fst) cur_env
+    -- Run doctest via cabal repl
+    exitCode <-
+        P.withCreateProcess
+            (P.proc cabal (mkArgs exePath args)){P.env = Just new_env}
+            (\_ _ _ p -> P.waitForProcess p)
+    exitWith exitCode
+  where
+    mkArgs exePath opts = baseOpts exePath ++ opts
+    baseOpts exePath = ["repl", "--with-compiler", exePath, "--with-hc-pkg", ghc_pkg]
```

Comment:
Will Cabal ensure that the project has been built before invoking external commands?

My default assumption is "no", and if that's the case, then I think we still need `--with-repl` to make this robust.

Reply:
Or given that we are in control now, maybe we are in a position to first invoke `cabal` with `-w ghc` and then  with `-w doctest`.

Reply:
> Will Cabal ensure that the project has been built before invoking external commands?

Yes, please give it a try.

Reply:
From what I tried, I think it doesn't.

Reply:
I just tried again:
- Delete `dist-newstyle` (or use --builddir`)
- Run `cabal doctest-9.8 -b QuickCheck`
It does run the tests.

Reply:
We are talking about two different things here.

This was my original question:

> Will Cabal ensure that the project has been built **before invoking external** commands?

The answer is no, it does not.

> It does run the tests.

All of this is done by your `cabal-doctest`, not by `cabal`.

Reply:
Ok. I do not understand why this is relevant then. Doctest run the tests with success, which seems the only thing that matters in my opinion.

Reply:
If you don't understand something, please always feel free to ask.

This is the relevant part from the README:

> Technically, `cabal build` is not necessary. `cabal repl --with-ghc=doctest` will build any dependencies as needed.  However, it's more robust to run `cabal build` first (specifically it is not a good idea to build  `ghc-paths` with `--with-ghc=doctest`).

Or in other words, before invoking `cabal repl`,, you have to run `cabal build`.

I'm giving this a stab right now, BTW.

Reply:
> If you don't understand something, please always feel free to ask.

If you want to ask something, please always feel free to give the _full_ **context**.

All examples I have seen only involved `cabal repl`, and my current PR just _works_ in my testing. So I believe you if you say it’s more robust, even if I do not know in what way: what can go wrong without `cabal build`?.

If you would have mentioned the note in your first comment it would have been clear and we would have both avoid losing time. This is my first contribution to this project, I do not know all the edge cases and it is not clear where the technical documentation of this one is.

Reply:
> I do not know in what way: what can go wrong without `cabal build`?

OK, I think that's the source of the misunderstanding.  I mentioned that before here:



https://github.com/haskell/cabal/pull/10057#issuecomment-2139384279

   > There's still the issue that when `doctest` acts as a proxy for GHC, then  the`ghc-paths` package is not complied correctly (not sure if any other package are affected).  For that reason a user should always do a clean build  first, before invoking `cabal repl --with-ghc=doctest`.

That's why I assumed you have the full picture.

That's also the reason why I kept pressing  for`--with-repl`, which would eliminate the problem + I think that's also the reason why Matthew [included this line in his script](https://github.com/mpickering/cabal-doctest-demo/blob/master/cabal-doctest#L20).

If you want to understand for what reason `ghc-paths` is not compiled correctly with `--with-ghc=doctest` then study https://github.com/simonmar/ghc-paths/blob/master/Setup.hs#L68.

Let's keep this constructive.

There are actual problems that need solving (e.g. [this line]( https://github.com/mpickering/cabal-doctest-demo/blob/master/cabal-doctest#L12), the `cabal exec`, is there for reason, and I think it makes things complicated, more on that later…).

Reply:
While I really appreciate that you now took time to give me the full picture, I do not feel I can be helpful here anymore.

I missed the important point of your [comment](https://github.com/haskell/cabal/pull/10057#issuecomment-2139384279) when I created this PR _2 months ago_, so it was unlikely I would understood the issue without some reference.

I am glad that at least my attempts motivate you to tackle #396.